### PR TITLE
feat: Create governance vc through governance VCS

### LINKED ISSUE #244: Create governance vc through governance VCS

### PR DESCRIPTION
**Pull request by fqutishat**
closes #244

Signed-off-by: Firas Qutishat <firas.qutishat@securekey.com>

### CODE REVIEW

File: test/bdd/fixtures/adapter-rest/docker-compose.yml [hunk 1]
```diff
@@ -53,7 +56,8 @@ services:
     ports:
       - 8070:8070
     entrypoint: ""
-    command:  /bin/sh -c "adapter-rest start"
+    # wait until governance vcs is up
+    command:  /bin/sh -c "sleep 20;adapter-rest start"
```
Comment (llorllale) [inline test/bdd/fixtures/adapter-rest/docker-compose.yml:60]: remove sleep

Comment (fqutishat) [reply test/bdd/fixtures/adapter-rest/docker-compose.yml:60]: need it

Comment (fqutishat) [reply test/bdd/fixtures/adapter-rest/docker-compose.yml:60]: need to make suer governance vcs is up

Comment (fqutishat) [reply test/bdd/fixtures/adapter-rest/docker-compose.yml:60]: I will add separate PR for retry

File: cmd/adapter-rest/startcmd/start.go [hunk 2]
```diff
@@ -628,13 +638,36 @@ func addRPHandlers(
 	return nil
 }
 
+func issueGovernanceCredential(governanceVCSURL string, rootCAs *x509.CertPool, store storage.Provider) error {
```
Comment (llorllale) [inline cmd/adapter-rest/startcmd/start.go:641]: Governance credentials are issued per RP and per Issuer, not per adapters.

The issuance of these credentials should be plugged in to the respective "create tenants" endpoints in the operations.

Also, the credential subject of these credentials should reference the public DIDs of the rps/issuers.

File: test/bdd/fixtures/adapter-rest/docker-compose.yml (hunk 1 above)
Comment (llorllale) [reply test/bdd/fixtures/adapter-rest/docker-compose.yml:60]: is there a TODO?

File: cmd/adapter-rest/startcmd/start.go (hunk 2 above)
Comment (fqutishat) [reply cmd/adapter-rest/startcmd/start.go:641]: please reference where is the create tenants endpoints

File: test/bdd/fixtures/adapter-rest/docker-compose.yml (hunk 1 above)
Comment (fqutishat) [reply test/bdd/fixtures/adapter-rest/docker-compose.yml:60]: I will add it

File: cmd/adapter-rest/startcmd/start.go (hunk 2 above)
Comment (llorllale) [reply cmd/adapter-rest/startcmd/start.go:641]: @fqutishat 

* issuer: https://github.com/trustbloc/edge-adapter/blob/d1e8c1705d913b72a15c91c842199c3ed123802c/pkg/restapi/issuer/operation/operations.go#L205
* rp: https://github.com/trustbloc/edge-adapter/blob/f41c373b400728be416f4f57a9b1faa14e1ca325/pkg/restapi/rp/operation/operations.go#L1105

Comment (fqutishat) [reply cmd/adapter-rest/startcmd/start.go:641]: @llorllale I will move the logic to  "create tenants" endpoints and add public dids

Comment (fqutishat) [reply cmd/adapter-rest/startcmd/start.go]: done

File: cmd/adapter-rest/startcmd/start.go [hunk 3]
```diff
@@ -635,6 +649,18 @@ func addIssuerHandlers(parameters *adapterRestParameters, ariesCtx ariespai.CtxP
...
+	if parameters.governanceVCSURL != "" {
+		var errNew error
+		// TODO add token parameter
+		governanceProvider, errNew = governance.New(parameters.governanceVCSURL, &tls.Config{RootCAs: rootCAs}, store,
+			map[string]string{})
+		if errNew != nil {
+			return errNew
+		}
+	}
+
```
Comment (rolsonquadras) [inline cmd/adapter-rest/startcmd/start.go:663]: May be we can combine initialization of governanceProvider for issuer/rp in one place.

File: pkg/governance/provider.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	req := &issueCredentialRequest{
+		Credential: []byte(fmt.Sprintf(governanceJSONLdDoc, didID)),
+	}
+
+	reqBytes, err := json.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+
+	endpointURL := fmt.Sprintf(issueCredentialURLFormat, p.governanceVCSUrl, governanceProfileName)
```
Comment (rolsonquadras) [inline pkg/governance/provider.go:127]: ideally profile name should be configurable (re: `governanceProfileName`). May be a TODO. Main reason for my thinking is that someone else already had registered with this name/id in vc services or if we want to change the governance profile to have other configuration (eg.DIDs etc.)

File: pkg/governance/provider.go [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return sendHTTPRequest(httpReq, p.httpClient, http.StatusCreated, p.requestTokens[vcsGovernanceRequestTokenName])
+}
+
+// GetCredential get governance credential.
+func (p *Provider) GetCredential() ([]byte, error) {
+	return p.store.Get(governanceVCKey)
```
Comment (rolsonquadras) [inline pkg/governance/provider.go:140]: Looks to be a single goverance vc for a deployment. If we are creating at profile/tenant level shouldn't we have multiple governance VCs ?

File: pkg/restapi/issuer/operation/operations.go [hunk 6]
```diff
@@ -253,6 +257,17 @@ func (o *Operation) createIssuerProfileHandler(rw http.ResponseWriter, req *http
 		return
 	}
 
+	if o.governanceProvider != nil {
```
Comment (llorllale) [inline pkg/restapi/issuer/operation/operations.go:260]: when are we going to make this mandatory?

File: pkg/governance/provider.go [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+   "geos": ["Canadian"],
+   "roles": ["accreditor"],
+   "privileges": [
+       {"name": "accredit", "uri": "https://example.com/accredit"}
+   ],
+  "duties": [
+       {"name": "safe-accredit", "uri": "https://example.com/responsible-accredit"}
+    ],
+   "define": [
+       {"name": "DID", "id": "%s"}
```
Comment (llorllale) [inline pkg/governance/provider.go:63]: the did ID will likely be placed somewhere else in the future, maybe credentialSubject.id

File: cmd/adapter-rest/startcmd/start.go (hunk 3 above)
Comment (fqutishat) [reply cmd/adapter-rest/startcmd/start.go]: done

File: pkg/governance/provider.go (hunk 4 above)
Comment (fqutishat) [reply pkg/governance/provider.go:127]: I add TODO

File: pkg/governance/provider.go (hunk 5 above)
Comment (fqutishat) [reply pkg/governance/provider.go]: done

**Pull request merged**